Add support for sending email via AWS SES

txase opened the pull request:
This code adds [AWS SES (Simple Email Service)](https://aws.amazon.com/ses/) as an additional (optionally enabled) email transport. It uses the same default AWS SDK config that the s3 file support uses.

> This is the second piece of functionality from the AWS Serverless RFC (#5591), following the recent merge of AWS S3 support for files (#5626). I plan to add docs for using all the AWS features together once each feature has been merged.

## Requirements for use
* An AWS account with SES enabled:
    * An email or domain identity for your sending address must be validated in SES, see [SES docs](https://docs.aws.amazon.com/ses/latest/dg/creating-identities.html) for details
    * The identity **must not** have a _Default Configuration Set_ (You can delete the default configuration set if you accidentally created it)
* Build vaultwarden with either the `ses` or `aws` feature
* Run vaultwarden with:
    * `USE_AWS_SES=true`
    * `SMTP_FROM=<sending email address>`
    * `AWS_PROFILE=<profile name>` (If not using a default AWS config profile. You may also use any other standard [AWS env vars](https://docs.aws.amazon.com/sdkref/latest/guide/environment-variables.html) to configure SDK credentials.)

septatrix commented:
Why does this need a custom implementation? AWS SES also provides an SMTP endpoint which you should already be able to use?

txase commented:
> Why does this need a custom implementation? AWS SES also provides an SMTP endpoint which you should already be able to use?

@septatrix The AWS SES SMTP endpoint only supports static credential authentication ([doc link](https://docs.aws.amazon.com/ses/latest/dg/smtp-credentials.html)). Static credentials add operational overhead (e.g. generating them the first time, keeping them safe, managing rotations, etc.), and are less secure relative to temporary session credentials.

This PR uses the AWS SES SDK that supports temporary session credentials via the SES API instead of the SMTP endpoint.

txase commented:
Note for reviewers: If #5917 is merged, I will update this PR to use the same approach to re-use the reqwest library for SES SDK calls.

txase commented:
I just pushed up a rebased and updated commit now that #5917 has been merged. This only change is that we now also use the built-in reqwest client implementation for AWS SDK calls, which significantly reduces the packages brought in as dependencies.

@dani-garcia @BlackDex: This PR is once again ready for review and merging, thanks!

BlackDex commented on src/mail.rs:
```diff
@@ -640,6 +680,15 @@ async fn send_with_selected_transport(email: Message) -> EmptyResult {
+            Err(e) => err!("Failed to send email", format!("Failed to send email using AWS SES: {e:?}")),
+        }
+
+        #[cfg(not(ses))]
+        err!("Failed to send email", "Failed to send email using AWS SES: `ses` feature is not enabled");
```
I would suggest to move this check to the `config.rs` file around line 927 where it starts with `if cfg._enable_smtp {`

There we also check if sendmail is enabled that the binary exists or not.
Checking if the feature is enabled there would prevent a system running thinking to use SES while that feature isn't enabled.

txase replied:
I just noticed that the PR already has a check at line 970-972 of config.rs:

```rust
...
        } else if cfg.use_aws_ses {
            #[cfg(not(ses))]
            err!("`USE_AWS_SES` is set, but the `ses` feature is not enabled in this build");
        } else {
...
```

This particular code is unreachable. I could change it to `unreachable!()` to make this more clear, or I could simply delete this line of code. What do you think makes more sense?

BlackDex replied:
Um, i think it can be removed in a way that this should probably never be called at all.
The validation is done during startup and admin config save, and thus should never ever be able to reach it.
And i see no point in having a redundant check here too.

sorry that i missed the `config.rs` part though.

txase replied:
I tried removing the code, and then realized that you have to return <something> from that `else if` branch. I then tried to figure out syntax to make that `else if` branch conditionally configured, but I don't think there's a simple syntax to do so without modifying the structure of the entire `send_with_selected_transport()` function.

Instead, I changed the `err!()` call to an `unreachable!()` call. This will at least make it clear to those reading the code that this branch should never be called upon in this circumstance.

Thoughts?

septatrix commented:
While implementing the trait does not offer any additional value it makes the code cleaner and easier to reason about. The only reason why it is currently possible to just implement and ad-hoc method cleanly is because for each transport the errors are handled differently with a giant match statement. If that code were to be refactored sometime in the future it would be significantly easier to do that with the trait approach

txase commented:
I agree that some of the code could be refactored to simplify if all the transports were lettre async transport implementations. In this PR my goal was to disturb as little existing code as possible. If the maintainers would rather a larger refactor to simplify the code (e.g. making the transport a Box around a dynamic implementation of the lettre async transport trait) as part of this PR, then I can look into it. That could also just as easily be done in a future PR.

txase committed:
mail: add AWS SES transport

Serverless AWS deployments should not need an SMTP service or
credentials just to send Vaultwarden mail. Allow mail delivery through
Amazon SES when USE_AWS_SES is enabled, while preserving the existing
SMTP and sendmail transports.

Add the ses feature and an aws umbrella feature. Keep mail configuration
validation strict by requiring SMTP_FROM for SES. Treat SES as a
configured mail transport for email 2FA.

Send MIME messages through the SESv2 SendEmail raw content path. Share
AWS SDK configuration with S3 so AWS clients use the same reqwest-backed
connector and credential loading behavior.
